feat: Single precision setting for backends

kratsg opened the pull request:
# Pull Request Description

- Resolves #911
- Resolves #912 

This will unify our backends further by allowing for precision configuration of ints and floats simultaneously, rather than configuring each one separately. Allowing a mix of precisions doesn't quite make sense and we should not encourage this behavior.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Unify setting precision for all tensorlib backends
* Provide 32/64 settings for all backends while keeping same defaults
```

matthewfeickert commented:
This looks fine, but we should have a PR that temporarily patches Issue #913 before this goes in. I can open that up.

kratsg commented:
> This looks fine, but we should have a PR that temporarily patches Issue #913 before this goes in. I've opened up PR #915 for this.

rebased and pushed and fixed.

matthewfeickert commented on src/pyhf/tensor/numpy_backend.py:
```diff
@@ -38,8 +38,12 @@ class numpy_backend(object):
+        self.dtypemap = {
+            'float': np.float64 if self.precision == '64b' else np.float32,
+            'int': np.int64 if self.precision == '64b' else np.int32,
+            'bool': np.bool_,
```
@kratsg one question: Why is this `np.bool_` (the underscore)?

matthewfeickert replied:
This is just a "Matthew doesn't know" question, not a "this is a problem" question.

kratsg replied:
https://stackoverflow.com/a/6205237/1532974 -- this is a numpy thing.

matthewfeickert replied:
Answer: https://stackoverflow.com/questions/6205020/numpy-types-with-underscore-int-float-etc/6205237#6205237
